src/utils.jl

```diff
@@ -97,3 +113,88 @@ function shuffle_split_stratified(
     end
     return inds1, inds2
 end
+
+_sample_dims(data::AbstractVector) = Colon()
+_sample_dims(data::AbstractArray{<:Any,3}) = (1, 2)
+
+"""
+    fold([f,] x::AbstractVector)
+    fold([f,] x::AbstractArray{<:Any,3}; dims=(1, 2))
+
+Compute the absolute deviation of `x` from `f(x)`, where `f` defaults to `Statistics.median`.
+
+`f` is generally a measure of central tendency. `dims` are the dimensions over which the
+estimator `f` reduces and are passed as kwargs to `f`.
+"""
+fold(f, data; dims=_sample_dims(data)) = abs.(data .- f(data; dims=dims))
+fold(data; kwargs...) = fold(Statistics.median, data; kwargs...)
+
+"""
+    rank_normalize(x::AbstractVector)
+    rank_normalize(x::AbstractArray{<:Any,3}; dims=(1, 2))
+
+Rank-normalize the inputs `x` along the dimensions `dims`.
+
+Rank-normalization proceeds by first ranking the inputs using "tied ranking"
+and then transforming the ranks to normal quantiles so that the result is standard
+normally distributed.
+"""
+function rank_normalize(x::AbstractArray{<:Any,3}; dims=(1, 2))
+    # TODO: can we avoid mapslices and the allocations here?
+    return mapslices(x; dims=dims) do xi
+        return reshape(rank_normalize(vec(xi)), size(xi))
+    end
+end
+function rank_normalize(x::AbstractVector)
+    values = similar(x, float(eltype(x)))
+    rank_normalize!(values, x)
+    return values
+end
+function rank_normalize!(values, x)
+    rank = StatsBase.tiedrank(x)
+    _normal_quantiles_from_ranks!(values, rank)
+    values .= StatsFuns.norminvcdf.(values)
```

Comment:
Without broadcasting?
```suggestion
    map!(StatsFuns.norminvcdf, values, values)
```